[ios_edit_menu]add native edit menu

hellohuanlin opened the pull request:
Support native edit menu on the engine side. 

Design doc: https://docs.google.com/document/d/16-8kn58h_oD902e7vPSh6W20aHRBJKyNOdSe5rbAe_g/edit?resourcekey=0-gVdJ3fbOybV70ZKeHU7fkQ&tab=t.0

*List which issues are fixed by this PR. You must list at least one issue.*

https://github.com/flutter/flutter/issues/103163

*If you had to change anything in the [flutter/tests] repo, include a link to the migration guide as per the [breaking change policy].*

## Pre-launch Checklist

- [x] I read the [Contributor Guide] and followed the process outlined there for submitting PRs.
- [x] I read the [Tree Hygiene] wiki page, which explains my responsibilities.
- [x] I read and followed the [Flutter Style Guide] and the [C++, Objective-C, Java style guides].
- [x] I listed at least one issue that this PR fixes in the description above.
- [x] I added new tests to check the change I am making or feature I am adding, or the PR is [test-exempt]. See [testing the engine] for instructions on writing and running engine tests.
- [x] I updated/added relevant documentation (doc comments with `///`).
- [x] I signed the [CLA].
- [x] All existing and new tests are passing.

If you need help, consider asking for advice on the #hackers-new channel on [Discord].

<!-- Links -->
[Contributor Guide]: https://github.com/flutter/flutter/wiki/Tree-hygiene#overview
[Tree Hygiene]: https://github.com/flutter/flutter/wiki/Tree-hygiene
[test-exempt]: https://github.com/flutter/flutter/wiki/Tree-hygiene#tests
[Flutter Style Guide]: https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo
[C++, Objective-C, Java style guides]: https://github.com/flutter/engine/blob/main/CONTRIBUTING.md#style
[testing the engine]: https://github.com/flutter/flutter/wiki/Testing-the-engine
[CLA]: https://cla.developers.google.com/
[flutter/tests]: https://github.com/flutter/tests
[breaking change policy]: https://github.com/flutter/flutter/wiki/Tree-hygiene#handling-breaking-changes
[Discord]: https://github.com/flutter/flutter/wiki/Chat

hellohuanlin commented on shell/platform/darwin/ios/framework/Source/FlutterTextInputPlugin.mm:
```diff
@@ -2513,6 +2537,15 @@ - (void)takeKeyboardScreenshotAndDisplay {
+  if (@available(iOS 16.0, *)) {
+    // TODO: get point and arrowDirection from args
+    CGPoint point = CGPointMake(100, 100);
+    UIEditMenuArrowDirection arrowDirection = UIEditMenuArrowDirectionUp;
```
@justinmc For Milestone 1, I just need a `point` and `arrowDirection` from the framework. 

Note that for arrow direction, `up` actually means the edit menu is BELOW the text field (triangle pointing upwards)

justinmc replied:
I believe global coordinates is the convention. I just checked and the Scribble feature sends global Offsets in its method channel calls.

The way that arrow position works in the framework is that it tries to fit above the anchor point, but if it's too big, then it goes below. So for the framework to provide arrowDirection are we able to hardcode the height of the native menu in the framework?

Also, if you for example purposely position the menu too close to the top and use arrowDirection `down`, does it hang off the screen?

hellohuanlin replied:
> Also, if you for example purposely position the menu too close to the top and use arrowDirection down, does it hang off the screen?

When the menu is positioned exceeding the bound, then it will be capped within bound. For example, below I set `y` to be `-1000`: 

![Simulator Screenshot - iPhone 15 Pro Max - 2024-01-26 at 13 10 29](https://github.com/flutter/engine/assets/41930132/43763b39-9188-4760-9124-d46ede454f70)


> The way that arrow position works in the framework is that it tries to fit above the anchor point, but if it's too big, then it goes below. So for the framework to provide arrowDirection are we able to hardcode the height of the native menu in the framework?

Interesting. Let me try to see if it is possible to do similar in the engine (e.g. get the size of the menu). If it's possible, then the framework can send 2 points (top and bottom), and engine can decide which one to use; 
If not possible, then we will have to use a "magic number" (either engine side or framework side)

justinmc replied:
Sounds good. On the framework we typically already have both an `anchorAbove` and `anchorBelow`, so it would be easy to send both.

hellohuanlin replied:
On the framework do you have the rectangle area? I found the new API that takes a rect and iOS will decide the arrow direction automatically: https://github.com/flutter/engine/pull/50095#discussion_r1468211520

justinmc commented on shell/platform/darwin/ios/framework/Source/FlutterTextInputPlugin.mm:
```diff
@@ -2513,6 +2537,15 @@ - (void)takeKeyboardScreenshotAndDisplay {
   _keyboardViewContainer.frame = _keyboardRect;
 }
 
+- (void)showNativeEditMenu:(NSDictionary*)dictionary {
```
Nit: Maybe "showSystemEditMenu" to avoid saying "native"?

justinmc replied:
Or even maybe showEditMenu is ok.

justinmc commented on shell/platform/darwin/ios/framework/Source/FlutterTextInputPlugin.mm:
```diff
@@ -1149,9 +1170,9 @@ - (BOOL)canPerformAction:(SEL)action withSender:(id)sender {
     // Forbid pasting images, memojis, or other non-string content.
-    return [UIPasteboard generalPasteboard].string != nil;
+    // hasStrings does not require paste permission
+    return [UIPasteboard.generalPasteboard hasStrings];
```
Good catch!

justinmc commented on shell/platform/darwin/ios/framework/Source/FlutterTextInputPlugin.mm:
```diff
@@ -856,9 +857,29 @@ - (instancetype)initWithOwner:(FlutterTextInputPlugin*)textInputPlugin {
+- (UIMenu*)editMenuInteraction:(UIEditMenuInteraction*)interaction
+          menuForConfiguration:(UIEditMenuConfiguration*)configuration
+              suggestedActions:(NSArray<UIMenuElement*>*)suggestedActions API_AVAILABLE(ios(16.0)) {
+  return [UIMenu menuWithChildren:suggestedActions];
```
So in future milestones the framework could pass UIMenuElements like this?

hellohuanlin replied:
Yes, the framework can send the items via `showNativeEditMenu` channel, and we can return it here.

hellohuanlin replied:
Custom actions (e.g. send email) can be a bit more involved (e.g. the engine would need to call back to the framework).

justinmc reviewed (commented):
The API looks good and simple so far. I think the one main thing to make sure of is how the framework will decide the arrow direction (commented below). Mainly, how we will keep the behavior of making sure the menu doesn't hang off the screen.

hellohuanlin commented on shell/platform/darwin/ios/framework/Source/FlutterTextInputPlugin.mm:
```diff
@@ -2513,6 +2541,14 @@ - (void)takeKeyboardScreenshotAndDisplay {
+- (void)showNativeEditMenu:(NSDictionary*)dictionary {
+  if (@available(iOS 16.0, *)) {
+    // TODO: get targetRect from args
+    CGRect targetRect = CGRectMake(100, 100, 100, 100);
```
@justinmc I found [an API](https://developer.apple.com/documentation/uikit/uieditmenuinteractiondelegate/3975821-editmenuinteraction?language=objc) that allows iOS to automatically decide the up/down arrow. Do you think you can provide a "target rect" from the framework side? The target rect is the rectangle that presents the menu.

justinmc replied:
Is that like a bounding box for the menu? Currently most menus use an `anchorAbove` and `anchorBelow`. One thing we could do is to create a full-width rectangle whose top and bottom are given by these two anchors. Or if we need something more specific, [here is where](https://github.com/flutter/flutter/blob/c576f0039d75716f04f82c8b58ca208c756b9835/packages/flutter/lib/src/widgets/text_selection_toolbar_anchors.dart#L27) we calculate the anchors. We should have most of the necessary metrics there, we just might need to rework the context menus to take rectangles.

hellohuanlin replied:
It's more like a bounding box for the selected text area (not the menu). Apple just use the rect's top and bottom points for `anchorAbove` and `anchorBelow` for the edit menu. I think the reason apple asks for the rect rather than the 2 points is for potential UI change that may show context menu on left or right.

hellohuanlin replied:
> Sounds like we are probably ok using the full-width solution mentioned above.

Ya a rect of any width will work. 

> we have it [here](https://github.com/flutter/flutter/blob/c576f0039d75716f04f82c8b58ca208c756b9835/packages/flutter/lib/src/widgets/text_selection_toolbar_anchors.dart#L46) and would just need to pipe it through.

If it's easy on your side, I prefer to take the whole rect, so we are feeding the right info into the API. But either way should work.

justinmc replied:
Sounds good, I'll play around with it when I work on the framework side.

hellohuanlin commented on shell/platform/darwin/ios/framework/Source/FlutterTextInputPlugin.mm:
```diff
@@ -2513,6 +2541,17 @@ - (void)takeKeyboardScreenshotAndDisplay {
+        [encodedTargetRect[@"x"] doubleValue], [encodedTargetRect[@"y"] doubleValue],
+        [encodedTargetRect[@"width"] doubleValue], [encodedTargetRect[@"height"] doubleValue]);
+    CGRect localTargetRect = [self.hostView convertRect:globalTargetRect toView:_activeView];
+    [_activeView showEditMenuWithTargetRect:localTargetRect];
```
This PR should be something ready for you to pull and try out. I tested with this: 

```
    _channel.invokeMethod<void>(
      'TextInput.showSystemContextMenu',
      <String, dynamic>{
        // global coords
        'target_rect': <String, dynamic>{
          'x': 0,
          'y': 100,
          'width': 100,
          'height': 50,
        }
      },
    );
```

hellohuanlin replied:
note that i also changed the channel name, since "edit" menu is an ios term.

hellohuanlin commented:
hi @justinmc this is ready for review. could you take a look when you get time? Thanks!

justinmc reviewed (commented):
This seems to work great. I was finally able to get my engine compiling and try this out and it looks really nice with the native fade in/out animation and everything.

I guess this is tied to the currently active text input? I noticed that the buttons seem to work out of the box, and the menu doesn't show unless an input is focused. Is the native API tied to a UITextInput instance? I wonder if this prevents the use case of creating this menu for some reason unrelated to text input (in future milestones, I guess).

Also, I noticed that while the buttons did perform the correct action automatically, they didn't seem to be aware of the current selection. See the screenshot below where the top menu is native and the bottom one is Flutter:



<img width="383" alt="Screenshot 2024-02-05 at 3 53 57 PM" src="https://github.com/flutter/engine/assets/389558/4abb1ce6-2485-42ed-ab74-d3c5ce0d140b">

Tomorrow I want to try sticking this into the real framework API and making sure that it works for a real world use case, I'll post back then.

hellohuanlin commented:
> I guess this is tied to the currently active text input? 

Yes. 

> I noticed that the buttons seem to work out of the box, and the menu doesn't show unless an input is focused. Is the native API tied to a UITextInput instance? 

Yes, the native API is tied to FlutterTextInputView, which is a UITextInput. In my PR I simply return the default `suggestedActions`, which contains actions based on current state of `FlutterTextInputView`. 

> I wonder if this prevents the use case of creating this menu for some reason unrelated to text input (in future milestones, I guess).

We should be able to add custom actions. 

> Also, I noticed that while the buttons did perform the correct action automatically, they didn't seem to be aware of the current selection. See the screenshot below where the top menu is native and the bottom one is Flutter

I think it's very likely because our `canPerformAction`'s implementation (https://github.com/flutter/engine/blob/307471f972f4993055e531013a635f916762234c/shell/platform/darwin/ios/framework/Source/FlutterTextInputPlugin.mm#L1169-L1180). We will need something like: 

```
if (action == @selector(copy:)) {
  return [self textInRange:_selectedTextRange].length > 0
}
```
Let me quickly try it out.

hellohuanlin commented:
> Also, I noticed that while the buttons did perform the correct action automatically, they didn't seem to be aware of the current selection. See the screenshot below where the top menu is native and the bottom one is Flutter

@justinmc Updated the PR. This should fix the issue you have: https://github.com/flutter/engine/pull/50095/files#diff-4c7b102c0690b8ec5e2212b079f5d69fe3f816c84e47ce94bc7bc89312f39e40R1179-R1181

The overview of iOS logic is: 

1. For milestone 1 (basic editing actions), we simply return `suggestedActions` which is passed into `menuForConfiguration` API's parameter. The `suggestedActions` is controlled by `canPerformAction` API. For example, if we return YES for `copy:` action and return NO for all other actions, then `suggestedActions` will only contain the `copy` button. Then after user tap on the button, the `copy:` function will be called. 

2. For milestone 2 (including custom actions, and other default actions like "lookup"), we have to manually append the items in `menuForConfiguration` API.

justinmc reviewed (commented):
That seemed to work to get me the correct menu buttons, thanks!

I've opened a draft PR with the code I'm using to play around with this: https://github.com/flutter/flutter/pull/143002. Just hacking it quickly and passing in the selection rect seems to work quite well:

![output](https://github.com/flutter/engine/assets/389558/05f066da-55a1-47b7-8e5a-40ed9037f3f8)

It even does the thing where it flips upside down when near the top of the screen:

<img width="360" alt="Screenshot 2024-02-06 at 11 50 21 AM" src="https://github.com/flutter/engine/assets/389558/411143e5-e598-4985-a70b-d8c6bbdd2c84">

Looks really promising.

### Implications of being tied to the text input connection

Because this menu is tied to the currently active text input connection in the engine, I'm going to say that it should be a part of EditableText, not AdaptiveTextSelectionToolbar on the framework side of things. The latter could be used by users for any sort of context menu, and they'll get unexpected behavior in that case, at least now for milestone 1.

In future milestones will it be possible to show a system menu for something other than text editing? Or actually, do we even need to? Say someone has built a drawing app, and long pressing on the drawing canvas shows a context menu with a "Paste" button. They want to handle the onPressed callback themselves, to do something like insert a text element into the drawing canvas with the contents of the clipboard. Is there no way around the clipboard access warning in that case, even in a native app?

### When to enable this feature

Currently in milestone 1, this feature should be enabled by a boolean parameter that is disabled by default, right? Do we need to worry about what happens if someone enables it on an unsupported platform or even an unsupported version of iOS? The framework doesn't know the iOS version if I'm not mistaken.

### Only other problem I'm seeing right now

The menu doesn't remove itself when I hot reload or even hot restart.

<img width="379" alt="Screenshot 2024-02-06 at 11 50 56 AM" src="https://github.com/flutter/engine/assets/389558/70610ffb-09a2-4a63-96b5-c0194498aad5">

justinmc commented:
I've got it working great on my end in my draft PR, including conditionally switching between system/Flutter menu based on supportsShowingSystemContextMenu 👍 . I want to play around with how the future milestones will work to make sure the API will support those well. I'll post back with more.

hellohuanlin commented:
> I want to play around with how the future milestones will work to make sure the API will support those well.

Sounds good. Let me know if you need anything from native side.

justinmc commented:
Let me know if this sounds right.

On the framework side I think we can do this by creating a new widget that switches between the system menu and the normal old AdaptiveTextSelectionToolbar based on what is supported. Let's call it AdaptiveAndSystemTextSelectionToolbar (and CupertinoAdaptiveAndSystemTextSelectionToolbar for Cupertino).

By default, for now, we wouldn't do anything with these new widgets. If a user wants to use system menus, they can pass this widget to contextMenuBuilder.

```dart
return TextField(
  contextMenuBuilder: (BuildContext context, EditableTextState editableTextState) {
    return AdaptiveAndSystemTextSelectionToolbar(
      editableTextState: editableTextState,
    );
  },
);
```

If using only Cupertino:


```dart
return CupertinoTextField(
  contextMenuBuilder: (BuildContext context, EditableTextState editableTextState) {
    return CupertinoAdaptiveAndSystemTextSelectionToolbar(
      editableTextState: editableTextState,
    );
  },
);
```

In the future, if we want to enable system menus by default, we can put these widgets in _defaultContextMenuBuilder.

### Milestone 1 - No customization
AdaptiveAndSystemTextSelectionToolbar creates the menu with the default buttons for text editing, both when native and when Flutter, with no ability to change anything. The only parameter is `editableTextState`, and it's required.


### Milestone 2 - Custom buttons
In addition to taking a targetRect, showSystemContextMenu will also take a list of buttonItems. Something like this:

```dart
SystemButtonItem({
  int id,
  String label,
  SystemButtonItemType type, // Used to indicate the paste button.
  // Anything else for style, etc.?
})
```

In order to tell the framework that a context menu button was pressed, a new platform channel method could be created:

```dart
ContextMenu.systemContextMenuButtonTapped({
  int id,
})
```

AdaptiveAndSystemTextSelectionToolbar could optionally take a list of these button items. Passing nothing gives the original behavior of using only the default buttons.

**Question:** What if someone just wants to add a custom button to the default buttons?

### Milestone 3 - Submenus
Add a `children` parameter of some sort to SystemButtonItem. The API of AdaptiveAndSystemTextSelectionToolbar stays the same.

hellohuanlin commented:
> By default, for now, we wouldn't do anything with these new widgets. If a user wants to use system menus, they can pass this widget to contextMenuBuilder.

It'd be great if we support secure paste out of box, without asking developers to migrate their code. The logic to enable it can be: 

```
if (
   // 1. run time flag, true for iOS 16, false otherwise
  supportsShowingSystemContextMenu
   // 2. compile time flag, true for developers opt'ed
   && optInSystemContextMenuBeta 
   // 3. only support text input for now. But could support to non-text widgets in the future
   && text_input_service.hasTextConnection) 
{
  // use system context menu
} else {
  // use flutter context menu
}
```

Then after milestone 2 is done, we can remove `optInSystemContextMenuBeta` compile time flag so that its enabled by default. 

> What if someone just wants to add a custom button to the default buttons?

This should be doable. Developers can append items to the list of default items (from `editableTextState.buttonItems`). Something like [this code sample](https://docs.google.com/document/d/16-8kn58h_oD902e7vPSh6W20aHRBJKyNOdSe5rbAe_g/edit?resourcekey=0-gVdJ3fbOybV70ZKeHU7fkQ&tab=t.0#heading=h.aw9t5anu3c1j).

justinmc reviewed (commented):
Another thing: The system context menu doesn't hide itself on a hot restart, even though the text field loses focus. Is there anything we can do about that? I don't think there's any way to be aware of that at the framework level... Maybe we could hack something if we had hideSystemContextMenu, like always calling that when a text input connection is opened. Or maybe it doesn't matter since it's only related to development.

hellohuanlin commented on shell/platform/darwin/ios/framework/Source/FlutterPlatformPlugin.mm:
```diff
@@ -149,11 +149,21 @@ - (void)handleMethodCall:(FlutterMethodCall*)call result:(FlutterResult)result {
   }
 }
 
+- (void)showSystemContextMenu:(NSDictionary*)args {
```
It should be easy to add. Let me try!

hellohuanlin replied:
Just added `ContextMenu.hideSystemContextMenu` (no params). Feel free to play around. 

> Or maybe it doesn't matter since it's only related to development and not a release build.

If it's easy for you, I'm open to just fit it here. But if it's hard, we can also leave it, and I can create an issue to track that (no strong opinion)

justinmc replied:
Ok I'll play around with the hot restart thing and see if I can fix it, otherwise we'll open an issue.

justinmc replied:
Thanks for the quick turnaround on hideSystemContextMenu. I'll try it out and post back.

hellohuanlin commented:
> @hellohuanlin One more thing that just came up. How does the keyboard shortcut cmd+v relate to secure paste? Do users of native apps see a warning when they paste that way, and what about Flutter apps?

I did a quick research on CMD+V, and wrote up a summary: https://github.com/flutter/flutter/issues/143999

justinmc commented:
Thanks for doing that!

justinmc commented:
@hellohuanlin Is there a way to be informed of when the context menu is hidden? I see that if I tap anywhere on the screen, the system hides the context menu, and Flutter is unaware. This is probably not vitally important, but it would help with some edge cases.

hellohuanlin commented:
> @hellohuanlin Is there a way to be informed of when the context menu is hidden? I see that if I tap anywhere on the screen, the system hides the context menu, and Flutter is unaware. This is probably not vitally important, but it would help with some edge cases.

Yes, it should be easy to add. Let me know if (or when) you need it.

hellohuanlin commented:
@justinmc Added the dismiss callback. 

From somewhere near:  https://github.com/flutter/flutter/blob/7a4c2465afe875355aaac3efee158084f13e6fc7/packages/flutter/lib/src/services/text_input.dart#L1916

Add a new case like: 
```
      case 'TextInputClient.onDismissSystemContextMenu':
        print('TextInputClient.onDismissSystemContextMenu is called');
        print('client is');
        print(args[0]);
```

justinmc commented:
Thank you for the quick turnaround! I'll try this out.

justinmc commented:
@hellohuanlin I'm starting to feel good about this based on my work over in https://github.com/flutter/flutter/pull/143002. I think one big question remains: Should the new platform channel calls be part of the text input channel, or should they be moved to their own new channel?

An argument in favor of keeping them part of text input is that currently, showing the menu requires an active text input connection. That's how it decides which buttons to show.

However, in future phases, we'd like to be able to show custom buttons in this menu. Users may want to show the menu in a non-text input related situation. For example, a drawing app may allow cut/copy/paste of drawing elements. If we want to support something like that, I think we'd have to move it off of the text input channel.

Do we ever want to support non-text-editing use cases? In the drawing app example, is secure paste still an issue when using the Flutter-rendered menu?

hellohuanlin commented:
> Should the new platform channel calls be part of the text input channel, or should they be moved to their own new channel?
> Do we ever want to support non-text-editing use cases? In the drawing app example, is secure paste still an issue when using the Flutter-rendered menu?

Good point. I think it's a good idea to change. I can use `platformChannel` instead (https://github.com/flutter/engine/blob/90d1ff04ae021f90a7fb1ce5584b5f3daa2feb5b/shell/platform/darwin/ios/framework/Source/FlutterEngine.mm#L619C3-L619C19). Let me know if you can access this channel in your code.

hellohuanlin commented:
@justinmc I just updated to use `platform` channel rather than `textInput` channel. I also renamed the method to `System. onDismissSystemContextMenu`. Let me know if any problem using it.

justinmc commented:
@hellohuanlin Can you fix the conflict here? I'm coming back to secure paste after finishing some other work finally.

hellohuanlin commented:
@justinmc welcome back! Updated the branch.

justinmc commented on shell/platform/darwin/ios/framework/Source/FlutterEngine.mm:
```diff
@@ -1082,6 +1082,12 @@ - (void)flutterTextInputView:(FlutterTextInputView*)textInputView
 
+- (void)flutterTextInputView:(FlutterTextInputView*)textInputView
+    willDismissEditMenuWithTextInputClient:(int)client {
+  [_platformChannel.get() invokeMethod:@"System.onDismissSystemContextMenu"
```
Should this be `ContextMenuonDismissSystemContextMenu` in order to align with `ContextMenu.showSystemContextMenu` etc.? Or is this the typical naming pattern?

hellohuanlin replied:
Renamed to `ContextMenu.onDismissSystemContextMenu`

justinmc reviewed (approved):
LGTM withe one last naming nit to consider.

Otherwise I have fully built the framework PR (https://github.com/flutter/flutter/pull/143002) on top of this and banged on it and everything seems to work as expected. @hellohuanlin can you give the framework PR a last look before we commit to this engine interface and merge? I'm holding off on finishing the tests until this is merged because I'm struggling to run them locally with a local engine, so ignore the sloppy tests for now.

hellohuanlin commented on shell/platform/darwin/ios/framework/Source/FlutterPlatformPlugin.mm:
```diff
@@ -149,11 +149,31 @@ - (void)handleMethodCall:(FlutterMethodCall*)call result:(FlutterResult)result {
   } else if ([method isEqualToString:@"Share.invoke"]) {
     [self showShareViewController:args];
     result(nil);
+  } else if ([method isEqualToString:@"ContextMenu.showSystemContextMenu"]) {
```
@justinmc Wanna double check before landing it:

For non-text input, If developers setup `contextMenuBuilder` to use system menu, does it call this method? Does it crash or does it just NO-OP?

justinmc replied:
Currently it calls it, even if there is no active text input connection. The result I see is that nothing happens. I wonder if there's anything we can do to prevent that unnecessary call, let me look on the framework side...

To be clear, at least the only way to do this now is to directly use SystemContextMenuController. It's not currently possible to use SystemContextMenu without an EditableText.

justinmc replied:
I don't think there's a good way for the framework to know whether or not there is an active text input connection at that point (in SystemContextMenuController). Maybe should the ContextMenu.showSystemContextMenu call indicate an error in the case where it doesn't succeed in showing the menu?

justinmc replied:
I'm adding some docs to the framework PR about this.

hellohuanlin replied:
> I don't think there's a good way for the framework to know whether or not there is an active text input connection at that point (in SystemContextMenuController). Maybe should the ContextMenu.showSystemContextMenu call indicate an error in the case where it doesn't succeed in showing the menu?

I think I can add a warning on the engine side.

hellohuanlin commented on shell/platform/darwin/ios/framework/Source/FlutterPlatformPlugin.mm:
```diff
@@ -149,11 +149,35 @@ - (void)handleMethodCall:(FlutterMethodCall*)call result:(FlutterResult)result {
+    BOOL shownEditMenu = [textInputPlugin showEditMenu:args];
+    if (!shownEditMenu) {
+      FML_LOG(ERROR) << "Only text input supports system context menu for now. See "
+                        "https://github.com/flutter/flutter/issues/143033.";
```
@justinmc added this warning if fails to show edit menu (e.g. when calling from non-text input).

justinmc replied:
Could you add this sentence if it's not too long?

"Ensure the system context menu is shown with an active text input connection."

That way it's more clear what someone needs to do to fix this, if it wasn't already obvious.

hellohuanlin commented on shell/platform/darwin/ios/framework/Source/FlutterTextInputPlugin.mm:
```diff
@@ -2511,6 +2549,23 @@ - (void)takeKeyboardScreenshotAndDisplay {
 }
 
+- (BOOL)showEditMenu:(NSDictionary*)args API_AVAILABLE(ios(16.0)) {
+  if (!self.activeView.isFirstResponder) {
```
@justinmc this is where I determine the menu is triggered by text input widgets vs non-text input widgets. 

I can also check `_textInputClient != 0`, but I have a feeling that there could be an edge case where `_textInputClient` is set, but the it is not yet in editing state. So I feel checking the first responder is safer.

matthew-carroll commented:
@hellohuanlin @justinmc - Is there any place where I can see how to use this functionality? We're trying to integrate it into super_editor.

I checked the design doc, but I can't tell which aspects were actually chosen vs which aspects were rejected.

I checked the file changes in this PR, but it looks all engine-side.

I also noticed above that two weeks ago this work may have been reverted? If that's the case, what's the plan to get it updated and rolled back in?

hellohuanlin commented:
@matthew-carroll the framework PR is here https://github.com/flutter/flutter/pull/143002

justinmc commented:
+1, ultimately it was relanded in https://github.com/flutter/flutter/pull/148265. Let me know how it goes integrating into super_editor.
